modules/dnn/src/model.cpp

```diff
@@ -0,0 +1,225 @@
+    };
+
+Model::Model(const Net& network) : Net(network), impl_(new Impl) {
+    impl_->outNames = getUnconnectedOutLayersNames();
+};
+
+Model& Model::setInputSize(const Size& size) {
+    impl_->size = size;
+    return *this;
+}
+
+Model& Model::setInputSize(int width, int height) {
+    impl_->size = Size(width, height);
+    return *this;
+}
+
+Model& Model::setInputMean(const Scalar& mean) {
+    impl_->mean = mean;
+    return *this;
+}
+
+Model& Model::setInputScale(float scale) {
+    impl_->scale = scale;
+    return *this;
+}
+
+Model& Model::setInputCrop(bool crop) {
+    impl_->crop = crop;
+    return *this;
+}
+
+Model& Model::setInputSwapRB(bool swapRB) {
+    impl_->swapRB = swapRB;
+    return *this;
+}
+
+void Model::predict(InputArray frame, OutputArrayOfArrays outs) {
+    std::vector<Mat> outputs;
+    outs.getMatVector(outputs);
+    impl_->predict(*this, frame.getMat(), outputs);
+}
+
+ClassificationModel::ClassificationModel(const std::string& model, const std::string& config)
+    : Model(model, config) {};
+
+ClassificationModel::ClassificationModel(const Net& network) : Model(network) {};
+
+std::vector<std::pair<int, float>> ClassificationModel::classify(InputArray frame) {
+    std::vector<Mat> outs;
+    impl_->predict(*this, frame.getMat(), outs);
+    CV_Assert(!outs.empty());
+
+    std::vector<std::pair<int, float>> predicts;
+    for (auto out : outs) {
+        double conf;
+        cv::Point maxLoc;
+        minMaxLoc(out, nullptr, &conf, nullptr, &maxLoc);
+        predicts.emplace_back(maxLoc.x, static_cast<float>(conf));
+    }
+    return predicts;
+}
+
+DetectionModel::DetectionModel(const std::string& model, const std::string& config)
+    : Model(model, config) {};
+
+DetectionModel::DetectionModel(const Net& network) : Model(network) {};
+
+void DetectionModel::detect(InputArray frame, CV_OUT std::vector<int>& classIds,
+                   CV_OUT std::vector<float>& confidences, CV_OUT std::vector<Rect2d>& boxes,
+                   float confThreshold, float nmsThreshold, bool absoluteCoords)
+{
+    int frameWidth  = frame.cols();
+    int frameHeight = frame.rows();
+
+    if (impl_->size.width <= 0) impl_->size.width = frame.cols();
+    if (impl_->size.height <= 0) impl_->size.height = frame.rows();
+
+    blobFromImage(frame, impl_->blob, 1.0, impl_->size, Scalar(), impl_->swapRB, impl_->crop, CV_8U);
+    setInput(impl_->blob, "", impl_->scale, impl_->mean);
+
+    if (getLayer(0)->outputNameToIndex("im_info") != -1) {  // Faster-RCNN or R-FCN
+        frameWidth  = impl_->size.width;
+        frameHeight = impl_->size.height;
+        Mat imInfo = (Mat_<float>(1, 3) << impl_->size.height, impl_->size.width, 1.6f);
+        setInput(imInfo, "im_info");
+    }
+
+    boxes.clear();
+    confidences.clear();
+    classIds.clear();
+
+    std::vector<Mat> detections;
+    impl_->outNames = getUnconnectedOutLayersNames();
+    forward(detections, impl_->outNames);
```

Comment:
We need to use `predict` method here.

Reply:
Move to Impl::predict?
```
if (getLayer(0)->outputNameToIndex("im_info") != -1) {  // Faster-RCNN or R-FCN
    frameWidth  = impl_->size.width;
    frameHeight = impl_->size.height;
    Mat imInfo = (Mat_<float>(1, 3) << impl_->size.height, impl_->size.width, 1.6f);
    setInput(imInfo, "im_info");
}
```
How to change frame size: `frameWidth  = impl_->size.width;  frameHeight = impl_->size.height;`?